api/handler.go

```diff
@@ -3,24 +3,168 @@ package api
 import (
 	"bytes"
 	"encoding/json"
-	"io"
-	"log"
+	"flag"
 	"net/http"
+	"os"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/gorilla/context"
+	"github.com/nvellon/hal"
+	"upper.io/db"
+	"upper.io/db/postgresql"
+
+	"github.com/thisissoon/deadpool/serializer"
 )
 
-func respond(w http.ResponseWriter, r *http.Request, status int, data interface{}) {
-	var buf bytes.Buffer
-	if err := json.NewEncoder(&buf).Encode(data); err != nil {
-		http.Error(w, err.Error(), http.StatusInternalServerError)
+var Database_uri string
+
+func init() {
+	databaseUrlUsage := `i.e.: postgres://postgres:mysecretpassword@postgres:5432/deadpool
+    or you can use env variable: DATABASE_URI`
+
+	flag.StringVar(&Database_uri, "d", os.Getenv("DATABASE_URI"), databaseUrlUsage)
+	flag.Parse()
+}
+
+type Response struct {
+	Count int
+	Total int
+}
+
+func (p Response) GetMap() hal.Entry {
+	return hal.Entry{
+		"count": p.Count,
+		"total": p.Total,
+	}
+}
+
+type ReportRest struct {
+	sess db.Database
+}
+
+func (self ReportRest) Get(w http.ResponseWriter, r *http.Request) {
+	col, err := self.sess.Collection("report")
+	if err != nil {
+		log.Fatal("Getting collection from db: ", err)
+		http.Error(w, "oops", http.StatusBadRequest)
 		return
 	}
 
-	w.WriteHeader(status)
-	if _, err := io.Copy(w, &buf); err != nil {
-		log.Println("respond: ", err)
+	limit := uint(20)
+	reps := col.Find().Limit(limit)
+	var reports []Report
+	reps.All(&reports)
+	c, _ := reps.Count()
+
+	halDoc := hal.NewResource(Response{Count: len(reports), Total: int(c)}, "")
+	for _, rep := range reports {
+		halDoc.Embed("reports", hal.NewResource(rep, ""))
 	}
+	doc, _ := json.Marshal(halDoc)
+	w.WriteHeader(http.StatusOK)
+	w.Write(doc)
 }
 
-func httpSaveReportHandler(w http.ResponseWriter, r *http.Request) {
-	respond(w, r, http.StatusOK, "# fasf ds?!")
+func (self ReportRest) Post(w http.ResponseWriter, r *http.Request) {
+	col, err := self.sess.Collection("report")
+	if err != nil {
+		log.Fatal("Getting collection from db: ", err)
+		http.Error(w, "oops", http.StatusBadRequest)
+		return
+	}
+
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(r.Body)
+	rep, err := serializer.UnmarshalReport(buf.Bytes())
+	if err != nil {
+		log.Error("Unmarshal report: ", err)
+		http.Error(w, "oops", 422)
+		return
+	}
+
+	rrr := Report{
```

Comment:
Rarrrrrrrrr? :stuck_out_tongue: 

Maybe a more descriptive variable name would be better here.